sale.sol

```diff
@@ -0,0 +1,153 @@
+  /// @param _spender The address of the account able to transfer the tokens
+  /// @return Amount of remaining tokens allowed to spent
+  function allowance(address _owner, address _spender) constant returns (uint256 remaining);
+
+  event Transfer(address indexed _from, address indexed _to, uint256 _value);
+  event Approval(address indexed _owner, address indexed _spender, uint256 _value);
+}
+
+contract Sale {
+  // We will pay funds to the association created.
+  address public beneficiary;
+
+  // Sale parameters
+  uint public fundingGoal;
+  uint public deadline;
+  uint public pricePerToken;
+  AbstractToken public rewardToken;
+
+  modifier afterDeadline() {
+    if (now < deadline) throw;
+    _;
+  }
+  
+  // Tracking variables
+  mapping(address => uint256) public contributionOf; // Track contributors
+  uint public amountRaised; //defaults to 0.
+  bool fundingGoalReached = false;
+  bool crowdsaleClosed = false;
+
+  // Events
+  event GoalReached(address beneficiary, uint amountRaised);
+  event FundsReceived(address backer, uint amount);
+  event FundsWithdrawal(address recipient, uint amount);
+  
+  function Crowdsale(
+    address beneficiary,
+    uint fundingGoalInEthers,
+    uint durationInMinutes,
+    uint etherCostPerToken,
+    AbstractToken rewardTokenAddress)
+  {
+    beneficiary = beneficiary;
+    fundingGoal = fundingGoalInEthers * 1 ether;
+    deadline = now + durationInMinutes * 1 minutes;
+    pricePerToken = etherCostPerToken * 1 ether;
+    rewardToken = AbstractToken(rewardTokenAddress);
+  }
+
+  // This function is called when the contract is paid ether.
+  function ()
+    payable
+  {
+    // Contribution possible?
+    // NOTE: what happens if contract is out of tokens?
+    if (crowdsaleClosed) throw;
+
+    uint amount = msg.value;
+    contributionOf[msg.sender] = amount;
+    amountRaised += amount;
+
+    // Send correct amount of tokens to the contributor.
+    // NOTE: Is this integer division?
+    rewardToken.transfer(msg.sender, amount / pricePerToken);
+    FundsReceived(msg.sender, amount);
+  }
+
+
+  function checkGoalReached()
+    afterDeadline
+  {
+    crowdsaleClosed = true;
+    if (amountRaised >= fundingGoal){
+      fundingGoalReached = true;
+      GoalReached(beneficiary, amountRaised);
+    }
+  }
+
+  // This function serves to functions. It allows contributors to safeWithdrawal
+  // their money if the sale fails to reach its goal and it allows
+  // the owner (or anyone) to transfer the funds raised to the beneficiary
+  // if the sale succeedded.
+  function withdraw()
+    afterDeadline
+  {
+    // This function cannot be run before validating whether the goal was reached,
+    // even if the deadline has passed.
+    if (!crowdsaleClosed) throw;
+
+    // Contributor withdrawing funds.
+    if (!fundingGoalReached) {
+      uint amount = contributionOf[msg.sender];
+      
+      if (amount > 0 && msg.sender.send(amount)) {
```

Comment:
Might be dangerous, because transaction also needs gas, therefore, people in the end would not be able to withdraw their funds. Also, do we really need withdrawals?